### PR TITLE
ELEMENTS-908: move to karma and open-wc for testing

### PR DESCRIPTION
**Pull request by nelsonsilva**

### CODE REVIEW

File: core/package.json [hunk 1]
```diff
@@ -11,23 +11,18 @@
   "author": "Nuxeo",
   "license": "Apache 2.0",
   "scripts": {
-    "test": "polymer test --npm --simpleOutput --expanded -l chrome,firefox",
-    "docs": "polymer analyze -i nuxeo-elements.js > analysis.json && polymer serve",
-    "coveralls": "cat coverage/lcov.info | coveralls"
+    "docs": "polymer analyze . > analysis.json && npm run serve",
```
Comment (nmpcunha) [inline core/package.json]: Is this running? We recently faced some issues with `polymer analyze` and we add the need to add `-i X` in order to workaround its automatic finding feature that seems to be buggy.

File: core/test/nuxeo-operation.test.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        '{"entity-type":"login","username":"Administrator"}',
+      ]);
+      server.respondWith('GET', '/api/v1/user/Administrator', [
+        200,
+        responseHeaders.json,
+        '{"entity-type":"user","username":"Administrator"}',
+      ]);
+    });
+
+    test('should retrieve something valid', async () => {
```
Comment (nmpcunha) [inline core/test/nuxeo-operation.test.js]: I know that it is not the scope of this ticket, but I'm not sure if I understand the purpose of this specific test...

File: core/test/test-helpers.js [hunk 3]
```diff
@@ -27,49 +29,67 @@ function timePasses(ms) {
...
-      }
-    });
-  });
+const flush = async () => {
+  utils.flush();
+  await timePasses(0);
+};
 
 function waitForEvent(el, event, times) {
   times = times || 1;
```
Comment (nmpcunha) [inline core/test/test-helpers.js]: Should we have a default value for the method param?

File: dataviz/package.json [hunk 4]
```diff
@@ -13,18 +13,14 @@
   "author": "Nuxeo",
   "license": "Apache 2.0",
   "scripts": {
-    "coveralls": "cat coverage/lcov.info | coveralls",
-    "test": "polymer test --npm --simpleOutput --expanded -l chrome,firefox",
-    "docs": "polymer analyze -i nuxeo-dataviz-elements.js > analysis.json && polymer serve"
+    "docs": "polymer analyze . > analysis.json && npm run serve",
```
Comment (nmpcunha) [inline dataviz/package.json]: Same question as above

File: ui/index.html [hunk 5]
```diff
@@ -5,10 +5,10 @@
 
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes" />
-    <script src="/node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
+    <script src="/components/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
 
-    <link rel="icon" href="favicon.ico" />
-    <script type="module" src="/node_modules/@polymer/iron-component-page/iron-component-page.js"></script>
+    <!--link rel="icon" href="favicon.ico" /-->
```
Comment (nmpcunha) [inline ui/index.html]: Should we remove it?

File: ui/package.json [hunk 6]
```diff
@@ -13,18 +13,14 @@
   "author": "Nuxeo",
   "license": "Apache 2.0",
   "scripts": {
-    "coveralls": "cat coverage/lcov.info | coveralls",
-    "test": "polymer test --npm --simpleOutput --expanded -l chrome,firefox",
-    "docs": "polymer analyze -i nuxeo-ui-elements.js > analysis.json && polymer serve"
+    "docs": "polymer analyze . > analysis.json && npm run serve",
```
Comment (nmpcunha) [inline ui/package.json]: Same as other `polymer analyze` instructions

File: core/package.json (hunk 1 above)
Comment (nelsonsilva) [reply core/package.json]: Note the added `.`. Polymer analyser should read our package.json and use the "main" as input.

File: core/test/nuxeo-operation.test.js (hunk 2 above)
Comment (nelsonsilva) [reply core/test/nuxeo-operation.test.js]: it's the basic "Hey, it works" use case, meaning it does not throw any exception. I did not really review the existing test tbh though, just migrated things over...

File: ui/index.html (hunk 5 above)
Comment (nelsonsilva) [reply ui/index.html]: Sorry, reverted the change. Thx 👍

File: core/test/test-helpers.js (hunk 3 above)
Comment (nelsonsilva) [reply core/test/test-helpers.js]: Done, thanks!

File: core/package.json (hunk 1 above)
Comment (nmpcunha) [reply core/package.json]: 😅
Should, indeed.

File: core/test/test-helpers.js [hunk 7]
```diff
@@ -14,62 +14,81 @@
...
-function login(server, loginFixture) {
-  server.respondWith('GET', '/dummy/json/cmis', [200, { 'Content-Type': 'application/json' }, '{}']);
+let server;
+
+async function login() {
+  server = sinon.fakeServer.create();
+  server.autoRespond = true;
 
-  server.respondWith('POST', '/dummy/api/v1/automation/login', [
+  server.respondWith('GET', '/json/cmis', [200, { 'Content-Type': 'application/json' }, '{}']);
```
Comment (Gabez0r) [inline core/test/test-helpers.js]: assuming we were exporting the responseHeaders:
```suggestion
  server.respondWith('GET', '/json/cmis', [200, responseHeaders.json, '{}']);
```

File: core/test/test-helpers.js [hunk 8]
```diff
@@ -14,62 +14,81 @@
...
+
+  server.respondWith('POST', '/api/v1/automation/login', [
     200,
     { 'Content-Type': 'application/json' },
```
Comment (Gabez0r) [inline core/test/test-helpers.js]:
```suggestion
    responseHeaders.json,
```
more of these below

File: core/test/test-helpers.js [hunk 9]
```diff
@@ -14,62 +14,81 @@
...
+ * This registers the fixture cleanup for Mocha's TDD interface
+ */
+teardown(() => {
+  if (server) {
+    server.restore();
+  }
+  fixtureCleanup();
+});
+
+export { timePasses, waitForEvent, waitChanged, login, fixture, fixtureCleanup, flush, html };
```
Comment (Gabez0r) [inline core/test/test-helpers.js]: export responseHeaders here

File: karma.conf.js [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+          statements: 90,
+          lines: 90,
+          branches: 90,
+          functions: 90,
+        },
+      },
+    },
+
+    client: {
+      mocha: {
```
[the remark this answers is not in the document]
Comment (nelsonsilva) [reply karma.conf.js]: I've changed to capture only browser console errors and fixed the few we had.

**Comment by Gabez0r** [comment]
👍

**Comment by guirenard** [comment]
:+1:

**Pull request closed without merging**